pkg/minikube/cluster/cluster_provisioner.go

```diff
@@ -0,0 +1,293 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This provisioner only for RHEL/CentOS iso which created to consume
+// with minishift
+
+package cluster
+
+import (
+	"bytes"
+	"bufio"
+	"errors"
+	"fmt"
+	"text/template"
+	"strings"
+	"path"
+	"os"
+	"syscall"
+
+	"github.com/docker/machine/libmachine/auth"
+	"github.com/docker/machine/libmachine/drivers"
+	"github.com/docker/machine/libmachine/engine"
+	"github.com/docker/machine/libmachine/log"
+	"github.com/docker/machine/libmachine/mcnutils"
+	"github.com/docker/machine/libmachine/provision/pkgaction"
+	"github.com/docker/machine/libmachine/swarm"
+	"github.com/docker/machine/libmachine/provision"
+
+	"golang.org/x/crypto/ssh/terminal"
+)
+
+var (
+	engineConfigTemplateRHEL = `[Unit]
+Description=Docker Application Container Engine
+After=network.target rc-local.service
+Requires=rhel-push-plugin.socket
+
+[Service]
+Type=notify
+ExecStart=/usr/bin/docker daemon -H tcp://0.0.0.0:{{.DockerPort}} -H unix:///var/run/docker.sock \
+           --authorization-plugin rhel-push-plugin
+           --storage-driver devicemapper --tlsverify --tlscacert {{.AuthOptions.CaCertRemotePath}} \
+           --tlscert {{.AuthOptions.ServerCertRemotePath}} --tlskey {{.AuthOptions.ServerKeyRemotePath}} \
+           {{ range .EngineOptions.Labels }}--label {{.}} {{ end }}{{ range .EngineOptions.InsecureRegistry }}--insecure-registry {{.}} {{ end }}{{ range .EngineOptions.RegistryMirror }}--registry-mirror {{.}} {{ end }}{{ range .EngineOptions.ArbitraryFlags }}--{{.}} {{ end }}
+ExecReload=/bin/kill -s HUP $MAINPID
+MountFlags=slave
+LimitNOFILE=infinity
+LimitNPROC=infinity
+LimitCORE=infinity
+TimeoutStartSec=0
+Delegate=yes
+KillMode=process
+Environment={{range .EngineOptions.Env}}{{ printf "%q" . }} {{end}}
+
+[Install]
+WantedBy=multi-user.target
+`
+	engineConfigTemplateCentOS = `[Unit]
+Description=Docker Application Container Engine
+After=network.target rc-local.service
+
+[Service]
+Type=notify
+ExecStart=/usr/bin/docker daemon -H tcp://0.0.0.0:{{.DockerPort}} -H unix:///var/run/docker.sock \
+           --storage-driver devicemapper --tlsverify --tlscacert {{.AuthOptions.CaCertRemotePath}} \
+           --tlscert {{.AuthOptions.ServerCertRemotePath}} --tlskey {{.AuthOptions.ServerKeyRemotePath}} \
+           {{ range .EngineOptions.Labels }}--label {{.}} {{ end }}{{ range .EngineOptions.InsecureRegistry }}--insecure-registry {{.}} {{ end }}{{ range .EngineOptions.RegistryMirror }}--registry-mirror {{.}} {{ end }}{{ range .EngineOptions.ArbitraryFlags }}--{{.}} {{ end }}
+ExecReload=/bin/kill -s HUP $MAINPID
+MountFlags=slave
+LimitNOFILE=infinity
+LimitNPROC=infinity
+LimitCORE=infinity
+TimeoutStartSec=0
+Delegate=yes
+KillMode=process
+Environment={{range .EngineOptions.Env}}{{ printf "%q" . }} {{end}}
+
+[Install]
+WantedBy=multi-user.target
+`
+)
+
+func init() {
+	provision.Register("minishift", &provision.RegisteredProvisioner{
```

Comment:
> Given that we already have a CentOS/RHEL provisioner, how is this ensuring that the right provisoner is run.

which provisioner to run is defined as per `osReleaseInfo.ID`. Existing rhel/centos provisioner only run when it find `rhel` or `centos` as part of ID. We are using `minishift` as ID for our images so it will make sure only this provisioner is going to run.

> I think RegisteredProvisioner is for adding provisioners for new OS types. Once you want to replace an existing one, you need a Detector as well.

We are not replacing any exiting provisioner here just adding one more to solver our problems.